### PR TITLE
Ensure that parallel cases under a non-parallel fixture only run with one another

### LINKED ISSUE #2143: Non-parallel fixture with parallel children runs in parallel with other fixtures
I'm breaking out this issue from the much larger complex of issues represented by #1474 because (1) it will be easier to solve and (2) it will require creating part of the solution for the larger issue.

Quoting from the above issue...

> Let's say you have a [NonParallelizable] test fixture. It's executing (by itself, of course) from the non-parallelizable queue. Now let's say that some of the test methods in that fixture are marked [Parallelizable]. As they are encountered, they will be added to the Parallelizable queue.

> Eventually, the dispatcher will exhaust the non-parallelizable queue and switech execution to the parallelizable queue and those test cases will get run. However, there may be other fixtures and their test cases on that queue and they will be run even though the original non-parallelizable fixture is still active.

[In fact, the problem can arise when __any__ type of non-parallel test has parallelizable children - for example, a non-parallelizable SetUpFixture with parallelizable TestFixtures under it. In this issue, however, let's deal with the simplest case of a non-parallelizable TestFixture containing parallelizable tests.]

What we would like to see happen here is this:
1. The fixture runs it's one-time setup in isolation, with no other fixtures running.
2. All the test cases run in isolation from other fixtures, even if they run in parallel.
3. The fixture runs it's one-time teardown, still in isolation.

In reality, only (1) and (3) happen correctly. That's because our notion of parallelism deals with what tests are __running__ and a fixture is only considered to be running when it's own code (not that of it's children) is executing.

In fact, we need to define a notion of a fixture being __active__ even if not running. It should be considered as active from the start of its one-time setup to the end of its one-time teardown. Our definition of parallelism has to switch from which tests are __running__ in parallel to which tests are __active__ in parallel. This is a big change, which is why I'm trying to start with a very limited test case.

As a first cut at the problem, I'm thinking of modifying how the parallel dispatcher is used.

Current Approach:
* The top level (composite) workitem, representing the assembly, is used to initialize the dispatcher.
* As each composite workitem is executed, it creates it's own children and dispatches them.
* The one-time teardown for a composite item is scheduled when all it's children have finished.
* Execution is complete when the top-level item finishes.

New Approach:
* ~~Prior to creating and dispatching work items, all fixtures will be located and identified as either parallelizable or non-parallelizable.~~
* ~~Non-parallel fixtures will be processed separately, one at a time, by adding the single fixture as the top-level item and re-starting the dispatcher shifts for each one of them.~~
* ~~Execution is complete when all queues, both fixture and work item, are empty.~~

>**Update:** The above approach complicated the code too much. Trying a different approach. See following comment.

Note that the above description __only__ solves the specific problem with test fixtures and test cases. For non-parallel setup fixtures, as described in #1474, multiple levels of recursion would be needed since SetUpFixtures may have other SetUpFixtures as well as TestFixtures as children. If the recursion looks simple, I'll try to fix that here as well. Otherwise a separate PR is needed.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #2143

### CODE REVIEW

**Comment by CharliePoole** [comment]
Not sure what happened here. The commits up to this point were merged through a different PR and GitHub says I closed it. It won't even let me reopen it, since there are no new commits. I'll add some more work and see if it re-opens.

**Comment by CharliePoole** [comment]
OK, I pushed a new commit and was able to get this to re-open.

In the new code, whenever any non-parallel fixture is started, a completely new set of queues is reserved for the children of that fixture. This prevents anything else executing while the fixture is active. Only tests that are children of that fixture will end up running. After that test is done, we go back to the original set of queues and continue working.

I'm not totally happy with the organization, but I don't see any better options...

1. The dispatcher has to control allocating and removing the queues, since it has to be done for all queues at once.
2. The change actually happens inside the WorkItemQueue, so that code is called by the dispatcher.
3. The TestWorker triggers the creation of the new queues, calling the dispatcher, because it can only happen after the work item is dequeued and before it is executed.
4. The WorkShift triggers  the removal of the queues when no longer needed, because that's where the end of a shift is detected.

I also wish I had a better way to test this than watching it in the debugger and examining the trace. Suggestions would be very welcome.

**Comment by CharliePoole** [comment]
I'm much happier with the code after the second commit. All control is now in ParallelWorkItemDispatcher, which uses private methods to save and restore the queues. WorkShift, WorkItemQueue and TestWorker are passive participants, which is as I think it should be.

@nunit/framework-team This is ready for review. If I can figure out how to do it, I'll add some tests. 🤢

**Comment by CharliePoole** [comment]
It's really annoying that an initialized but unused field doesn't give a warning in VS but does so on Travis.

**Comment by jnm2** [comment]
@CharliePoole
> It's really annoying that an initialized but unused field doesn't give a warning in VS but does so on Travis.

The only explanation for this that comes to mind is conditional compilation or debug- vs release-specific settings in the csproj?

**Comment by CharliePoole** [comment]
Nope. I do both locally and the same difference exists between Appveyor and Travis.

**Comment by tom-dudley** [comment]
Looks like the XBuild version of CS0413 doesn't match the 'spec' used by msbuild: [https://docs.microsoft.com/en-us/dotnet/articles/csharp/misc/cs0414](https://docs.microsoft.com/en-us/dotnet/articles/csharp/misc/cs0414) which only applies to constant values.

The XBuild version seems a bit saner to me..

**Comment by CharliePoole** [comment]
@tom-dudley Well, it's good to know anyway. 😄

**Comment by CharliePoole** [comment]
While waiting, I did some refactoring, which will help with the next bit of work. I'm adding it to this PR since nobody has reviewed it yet.

The new commit takes the creation of child work items completely away from the dispatcher and composite work item. All work items are now created before the dispatcher is invoked. This simplifies CompositeWorkItem by a small amount but, more important, it gives the dispatcher access to the full tree of work items in order to later implement dependencies between them.

Even so, I hope somebody will review this soon @nunit/framework-team !!!

**Comment by CharliePoole** [comment]
BTW, we are unable to build on travis using Mono 5.0, so I turned the clock back to 4.8.1 and put latest in the allowed failures for now.

**Comment by CharliePoole** [comment]
Turns out that "latest" is still 4.8.1 on OSX.

**Comment by CharliePoole** [comment]
@nunit/framework-team @nunit/core-team 
Now the weekend is over, I hope one of you will review this. :smile:

File: src/NUnitFramework/tests/Attributes/TestOrderAttributeTests.cs [hunk 1]
```diff
@@ -1,11 +1,9 @@
-﻿using System;
-using NUnit.Framework.Internal;
+﻿using NUnit.Framework.Internal;
 using NUnit.Framework.Internal.Execution;
 using NUnit.TestData;
-using NUnit.TestData.TestCaseSourceAttributeFixture;
 using NUnit.TestUtilities;
 
-namespace NUnit.Framework.Tests.Attributes
+namespace NUnit.Framework.Attributes
```
Comment (rprouse) [inline src/NUnitFramework/tests/Attributes/TestOrderAttributeTests.cs]: Curious, why the change in namespace?

File: src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            _dispatcher = new ParallelWorkItemDispatcher(LEVEL_OF_PARALLELISM);
+        }
+
+        [Test]
+        public void ConstructorCreatesQueues()
+        {
+            var queues = new List<WorkItemQueue>(_dispatcher.Queues);
+
+            Assert.That(List.Map(queues).Property("Name"),
```
Comment (rprouse) [inline src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs]: Minor, but assuming we have LINQ, these two lines could be simplified down to `_dispatcher.Queues.Select(q => q.Name)` and then it would be refactor safe.

File: src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                Is.EquivalentTo(new[] { "ParallelQueue", "NonParallelQueue", "ParallelSTAQueue", "NonParallelSTAQueue" }));
+        }
+
+        [Test]
+        public void ConstructorCreatesShifts()
+        {
+            var shifts = new List<WorkShift>(_dispatcher.Shifts);
+
+            Assert.That(shifts, Is.Unique);
+            Assert.That(shifts.Count(), Is.EqualTo(3));
```
Comment (rprouse) [inline src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs]: shifts is a List, why use Count() over Count? `Assert.That(shifts, Has.Count.EqualTo(3));`?

File: src/NUnitFramework/tests/Attributes/TestOrderAttributeTests.cs (hunk 1 above)
Comment (CharliePoole) [inline src/NUnitFramework/tests/Attributes/TestOrderAttributeTests.cs]: Most attribute tests are in NUnit.Framework.Attributes, with a few in NUnit.Framework.Tests.Attributes. I considered a mass change but didn't want to complicate this PR any further.

File: src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs (hunk 2 above)
Comment (CharliePoole) [inline src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs]: Good point... after years of not having it, I tend to forget about it.

File: src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs (hunk 3 above)
Comment (CharliePoole) [inline src/NUnitFramework/tests/Internal/Execution/ParallelWorkItemDispatcherTests.cs]: Copy / paste :smile:

**Comment by rprouse** [comment]
I've gone ahead and merged since nobody else has had a chance to code review.
